packages/flutter/lib/src/material/button_bar.dart

```diff
@@ -172,3 +181,183 @@ class ButtonBar extends StatelessWidget {
+/// the widget, it then aligns its buttons in a column. The key difference here
+/// is that the [MainAxisAlignment] will then be treated as a
+/// cross-axis/horizontal alignment. For example, if the buttons overflow and
+/// [ButtonBar.alignment] was set to [MainAxisAligment.start], the buttons would
+/// align to the horizontal start of the button bar.
+class RenderButtonBarRow extends RenderFlex {
+  /// Creates a button bar that attempts to display in a row, but displays in
+  /// a column if there is insufficient horizontal space.
+  RenderButtonBarRow({
+    List<RenderBox> children,
+    Axis direction = Axis.horizontal,
+    MainAxisSize mainAxisSize = MainAxisSize.max,
+    MainAxisAlignment mainAxisAlignment = MainAxisAlignment.start,
+    CrossAxisAlignment crossAxisAlignment = CrossAxisAlignment.center,
+    TextDirection textDirection,
+    VerticalDirection verticalDirection = VerticalDirection.down,
+    TextBaseline textBaseline,
+  }) : super(
+    children: children,
+    direction: direction,
+    mainAxisSize: mainAxisSize,
+    mainAxisAlignment: mainAxisAlignment,
+    crossAxisAlignment: crossAxisAlignment,
+    textDirection: textDirection,
+    verticalDirection: verticalDirection,
+    textBaseline: textBaseline,
+  );
+
+  bool _hasCheckedLayoutWidth = false;
+
+  @override
+  BoxConstraints get constraints {
+    if (_hasCheckedLayoutWidth)
+      return super.constraints;
+    return super.constraints.copyWith(maxWidth: double.infinity);
+  }
+
+  @override
+  void performLayout() {
+    // Set check layout width to false in reload or update cases.
+    _hasCheckedLayoutWidth = false;
+
+    // Perform layout to ensure that button bar knows how wide it would
+    // ideally want to be.
+    super.performLayout();
+    _hasCheckedLayoutWidth = true;
+
+    // If the button bar is constrained by width and it overflows, set the
+    // buttons to align vertically. Otherwise, lay out the button bar
+    // horizontally.
+    if (size.width > constraints.maxWidth) {
+      RenderBox child = firstChild;
+      double currentHeight = 0.0;
+      while (child != null) {
+        final FlexParentData childParentData = child.parentData;
+        assert(child.parentData == childParentData);
+
+        // Lay out the child with the button bar's original constraints, but
+        // with minimum width set to zero.
+        child.layout(constraints.copyWith(minWidth: 0.0), parentUsesSize: true);
+
+        // Set the cross axis alignment for the column to match the main axis
+        // alignment for a row. For [MainAxisAligment.spaceAround],
+        // [MainAxisAligment.spaceBetween] and [MainAxisAlignment.spaceEvenly]
+        // cases, use [MainAxisAligmnent.start].
+        switch (textDirection) {
+          case TextDirection.ltr:
+            switch (mainAxisAlignment) {
+              case MainAxisAlignment.center:
+                final double midpoint = constraints.maxWidth / 2.0 - child.size.width / 2.0;
```

Comment:
Really minor style thing, but it might be better to express this as:

```
(constraints.maxWidth - child.size.width) / 2.0
```

As it is probably less likely to loose precision with one division instead of two. Might be faster too.